tests/unit_tests/test_analytics/test_sentry.py

```diff
@@ -0,0 +1,380 @@
+WANDB_DSN_URL_FORMAT = SENTRY_DSN_FORMAT.format(
+    key=INTERNAL_SENTRY_DSN_KEY,
+    address=API_SERVER_ADDRESS,
+    project=INTERNAL_SENTRY_PROJECT,
+)
+EXTERNAL_DSN_URL_FORMAT = SENTRY_DSN_FORMAT.format(
+    key=EXTERNAL_SENTRY_DSN_KEY,
+    address=API_SERVER_ADDRESS,
+    project=EXTERNAL_SENTRY_PROJECT,
+)
+
+
+@pytest.fixture(scope="module")
+def relay():
+    relay_server = MetricRelayServer()
+    relay_server.start()
+
+    return relay_server
+
+
+class MetricRelayServer:
+    events = {}
+
+    def __init__(
+        self,
+    ) -> None:
+        self.is_running = False
+        self.port = self._get_free_port()
+        self.app = flask.Flask(__name__)
+        self.app.add_url_rule(
+            rule="/api/<project_id>/envelope/",
+            methods=["POST"],
+            view_func=self.sentry,
+        )
+        self.app.add_url_rule(
+            rule="/ping/",
+            methods=["GET"],
+            view_func=self.ping,
+        )
+        self.relay_url = f"http://127.0.0.1:{self.port}"
+
+        self._sentry_responses = []
+
+    def start(self) -> None:
+        # run server in a separate thread
+        relay_server_thread = threading.Thread(
+            target=self.app.run,
+            kwargs={
+                "port": self.port,
+            },
+            daemon=True,
+        )
+        relay_server_thread.start()
+
+        # wait for server to start
+        while not self.is_running:
+            try:
+                requests.get(f"{self.relay_url}/ping/")
+                self.is_running = True
+            except BaseException:
+                continue
+
+        self.is_running = True
+
+    @staticmethod
+    def _get_free_port() -> int:
+        sock = socket.socket()
+        sock.bind(("127.0.0.1", 0))
+        _, port = sock.getsockname()
+        return port
+
+    def extract_sentry_message_details(self, event_data):
+        decompressed_data = zlib.decompress(event_data, 16 + zlib.MAX_WBITS)
+        envelope = Envelope.deserialize(decompressed_data)
+        return envelope
+
+    def sentry(self, project_id):
+        decompressed_data = zlib.decompress(request.get_data(), 16 + zlib.MAX_WBITS)
+        envelope = Envelope.deserialize(decompressed_data)
+        self.events[envelope.headers["event_id"]] = {
+            "envelope": envelope,
+            "project_id": project_id,
+        }
+        return "OK"
+
+    def ping(self):
+        return "OK"
+
+
+def wait_for_events(relay, event_ids, timeout=5):
+    start_time = time.time()
+    end_time = start_time + timeout
+
+    while (
+        not all(event_id in relay.events for event_id in event_ids)
+        and time.time() < end_time
+    ):
+        time.sleep(0.1)
+
+    for event_id in event_ids:
+        assert event_id in relay.events
+
+
+# Helps in checking that no data is leaked between wandb sentry events and client sentry events
+def assert_wandb_and_client_events(wandb_envelope, client_envelope):
```

Comment:
maybe instead of this function, let's add some logic around envelope (maybe another class?) that will return the relevant information, like: `public_key`, `payload`, `tags`, `project_id`

Reply:
still don't like this function :)

Reply:
we could add methods/or operators on the response class